server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java

```diff
@@ -148,50 +149,68 @@ public JoinValidationService(
     }
 
     public void validateJoin(DiscoveryNode discoveryNode, ActionListener<Void> listener) {
-        if (discoveryNode.getVersion().onOrAfter(Version.V_8_3_0)) {
+        // This node isn't in the cluster yet so ClusterState#getMinTransportVersion() doesn't apply, we must obtain a specific connection
+        // so we can check its transport version to decide how to proceed.
+
+        final Transport.Connection connection;
```

Comment:
Nit: who is going to close the connection?
The behaviour should be the same as before, as the connection was opened but not closed in the original location in doRun as well, so I suppose it's not a problem.

Reply:
We are not opening the connection here, just getting a connection that someone else is responsible for opening. Its lifecycle is managed here:

https://github.com/elastic/elasticsearch/blob/7e753fb021cfe7123018b183b0d0014282a2137d/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java#L634-L636

Reply:
Thanks for the code pointer!
Yeah, I should have said "obtained", not "opened" :)